### PR TITLE
Drop MyISAM support

### PR DESCRIPTION
**Pull request by AdrienClairembault**
## Checklist before requesting a review

- [X] I have performed a self-review of my code.
- [X] I have added tests that prove my fix is effective or that my feature works.

## Description

Drop MysIsam support.

## Screenshots

CLI:
image

UI:
image

### CODE REVIEW

**Comment by AdrienClairembault** [comment]
Test failure is unrelated, I'll rebase once main is fixed.

File: src/DBConnection.php [hunk 1]
```diff
@@ -110,7 +103,6 @@ public static function createMainConfig(
         bool $use_timezones = false,
         bool $log_deprecation_warnings = false,
         bool $use_utf8mb4 = false,
-        bool $allow_myisam = true,
```
Comment (cedric-anne) [inline src/DBConnection.php:113]: Please add a changelog entry:
```
`DBConnection::createMainConfig()` signature changed. The `$allow_myisam` parameter has been removed.
```

File: src/DBConnection.php [hunk 2]
```diff
@@ -254,7 +242,6 @@ public static function createSlaveConnectionFile(
...
```
Comment (cedric-anne) [inline src/DBConnection.php:257]: Same here.

File: src/DBmysql.php [hunk 3]
```diff
@@ -150,14 +150,6 @@ class DBmysql
      */
     public $use_utf8mb4 = false;
 
-    /**
-     * Determine if MyISAM engine usage should be allowed for tables creation/altering operations.
-     * Defaults to true to keep backward compatibility with old DB.
-     *
-     * @var bool
-     */
-    public $allow_myisam = true;
```
Comment (cedric-anne) [inline src/DBmysql.php:159]: Please add a changelog entry for this removal.

File: src/Glpi/System/Requirement/TablesEngine.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+namespace Glpi\System\Requirement;
+
+/**
+ * @since 9.5.0
```
Comment (cedric-anne) [inline src/Glpi/System/Requirement/TablesEngine.php:39]:
```suggestion
 * @since 11.0.0
```

File: src/Glpi/System/Requirement/TablesEngine.php [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+ */
+final class TablesEngine extends AbstractRequirement
```
Comment (cedric-anne) [inline src/Glpi/System/Requirement/TablesEngine.php:41]:
```suggestion
final class DatabaseTablesEngine extends AbstractRequirement
```

File: src/Glpi/System/Requirement/TablesEngine.php [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+     * DB instance.
+     *
+     * @var \DBmysql
+     */
+    private $db;
+
+    public function __construct(\DBmysql $db)
+    {
+        parent::__construct(
+            __('DB tables engine')
```
Comment (cedric-anne) [inline src/Glpi/System/Requirement/TablesEngine.php:53]:
```suggestion
            __('Database tables engine')
```

File: src/Glpi/System/Requirement/TablesEngine.php [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            );
+        }
+
+        // List each invalid table
+        foreach ($tables as $table) {
+            $this->validation_messages[] = sprintf(
+                __('The "%1$s" table does not have the required InnoDB engine.'),
+                $table['TABLE_NAME']
+            );
+        }
```
Comment (cedric-anne) [inline src/Glpi/System/Requirement/TablesEngine.php:79]: We should not list all the tables. Someone that has not yet run the migration would have a list of 300+ tables displayed.
Also, the `Please run the command` part should probably be added after the explaination of the validation failure.

I propose the following message:
> The database contains %d tables using the unsupported MyISAM engine. Please run the "%s" command to migrate them to the InnoDB engine.

File: src/Glpi/System/RequirementsManager.php [hunk 8]
```diff
@@ -200,6 +197,7 @@ function ($directory) {
         );
 
         if ($db instanceof \DBmysql) {
+            $requirements[] = new DbEngine($db);
```
Comment (cedric-anne) [inline src/Glpi/System/RequirementsManager.php:200]: It should not be moved here. The display order in the UI correspond to the declaration order here.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Comment by trasher** [comment]
Please fix tests

**Comment by cedric-anne** [comment]
Thinking about it again, I realize that completely blocking the creation of MyISAM tables is going to be painful to manage in the plugins. This PR was proposed to fix an issue on a GLPI table, I guess the `DatabaseTableEngine` requirement could only check GLPI core tables (the `DBmysql::getMyIsamTables()` have a boolean parameter to exclude plugins tables), and we could keep the previous behaviour (warning when a MyISAM table is created + warnings on the central page).

Anyway, I have no solution that could permit to prevent changes on the old migrations in GLPI itself.

**Comment by AdrienClairembault** [comment]
Is it really that bad for plugins ? This can be fixed at once by a simple search and replace.
It feels like a minor issues compared to the whole auto escaping changes that will require a million times more more work for plugins maintainers.

**Comment by trasher** [comment]
Fixing auto escaping is absolutely not comparable.

Are you arguing there already are breaking changes to introduce another one?
We have to think about how we can do the best to ease migration; and force changes only when it's really not possible (this is totally acceptable anyway).

**Comment by AdrienClairembault** [comment]
> Are you arguing there already are breaking changes to introduce another one?

I am arguing that it takes 5 second to fix and that it wont be the painful part of migrating plugins to GLPI 11.

Forcing people to remove references to an engine that's been outdated for almost 15 years doesn't seem unreasonable to me.

**Comment by AdrienClairembault** [comment]
Rebased to fix conflicts, only "Revert warning" commit was added.

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**